### PR TITLE
More proper MPI etiquette, overlapping halo communication, and better scaling benchmarks

### PR DESCRIPTION
**Pull request by ali-ramadhan**
This PR adds MPI waits after sending and receiving all the halos and overlaps halo communication (thanks @christophernhill!).

I was initially worried that overlapping halo communication would cause the corner halos to become wrong (I'm not even sure if the corner halos should be filled via communication?). But running the MPI example the movie looked fine so I'll keep the overlapping for now.

Scaling benchmarks are better now! Results seem quite machine-dependent but on Satori for weak scaling I'm seeing >90% efficiency up to 16 ranks then ~85% efficiency on 32 ranks. Strong scaling seems noisier/a bit worse for some reason (including a 104% efficiency hit haha) but definitely better than before.

# Weak scaling (shallow water)

```
                            Shallow water model weak scaling benchmark
┌──────────────┬─────────┬─────────┬─────────┬─────────┬─────────┬────────────┬────────┬─────────┐
│         size │   ranks │     min │  median │    mean │     max │     memory │ allocs │ samples │
├──────────────┼─────────┼─────────┼─────────┼─────────┼─────────┼────────────┼────────┼─────────┤
│  (8192, 256) │  (1, 1) │ 1.188 s │ 1.190 s │ 1.190 s │ 1.193 s │ 416.81 KiB │   2665 │       5 │
│  (8192, 512) │  (1, 2) │ 1.187 s │ 1.192 s │ 1.192 s │ 1.200 s │ 408.80 KiB │   3178 │      10 │
│ (8192, 1024) │  (1, 4) │ 1.204 s │ 1.206 s │ 1.206 s │ 1.212 s │ 408.80 KiB │   3178 │      20 │
│ (8192, 2048) │  (1, 8) │ 1.220 s │ 1.223 s │ 1.223 s │ 1.230 s │ 408.80 KiB │   3178 │      40 │
│ (8192, 4096) │ (1, 16) │ 1.281 s │ 1.283 s │ 1.284 s │ 1.291 s │ 408.80 KiB │   3178 │      64 │
│ (8192, 8192) │ (1, 32) │ 1.347 s │ 1.417 s │ 1.424 s │ 1.497 s │ 408.80 KiB │   3178 │     128 │
└──────────────┴─────────┴─────────┴─────────┴─────────┴─────────┴────────────┴────────┴─────────┘
```

```
               Shallow water model weak scaling speedup
┌──────────────┬─────────┬──────────┬────────────┬──────────┬────────┐
│         size │   ranks │ slowdown │ efficiency │   memory │ allocs │
├──────────────┼─────────┼──────────┼────────────┼──────────┼────────┤
│  (8192, 256) │  (1, 1) │      1.0 │        1.0 │      1.0 │    1.0 │
│  (8192, 512) │  (1, 2) │  1.00165 │   0.998349 │ 0.980769 │ 1.1925 │
│ (8192, 1024) │  (1, 4) │  1.01309 │   0.987079 │ 0.980769 │ 1.1925 │
│ (8192, 2048) │  (1, 8) │  1.02721 │   0.973514 │ 0.980769 │ 1.1925 │
│ (8192, 4096) │ (1, 16) │  1.07822 │   0.927452 │ 0.980769 │ 1.1925 │
│ (8192, 8192) │ (1, 32) │  1.19049 │    0.83999 │ 0.980769 │ 1.1925 │
└──────────────┴─────────┴──────────┴────────────┴──────────┴────────┘
```

# Strong scaling (shallow water)

```
                             Shallow water model strong scaling benchmark
┌──────────────┬─────────┬──────────┬──────────┬──────────┬──────────┬────────────┬────────┬─────────┐
│         size │   ranks │      min │   median │     mean │      max │     memory │ allocs │ samples │
├──────────────┼─────────┼──────────┼──────────┼──────────┼──────────┼────────────┼────────┼─────────┤
│ (8192, 8192) │  (1, 1) │ 37.687 s │ 37.687 s │ 37.687 s │ 37.687 s │ 417.50 KiB │   2709 │       1 │
│ (8192, 8192) │  (1, 2) │ 19.278 s │ 19.280 s │ 19.280 s │ 19.282 s │ 409.48 KiB │   3222 │       2 │
│ (8192, 8192) │  (1, 4) │  9.027 s │  9.030 s │  9.029 s │  9.031 s │ 409.48 KiB │   3222 │       4 │
│ (8192, 8192) │  (1, 8) │  5.499 s │  5.514 s │  5.513 s │  5.520 s │ 409.48 KiB │   3222 │       8 │
│ (8192, 8192) │ (1, 16) │  2.587 s │  2.598 s │  2.599 s │  2.608 s │ 409.48 KiB │   3222 │      32 │
│ (8192, 8192) │ (1, 32) │  1.355 s │  1.437 s │  1.438 s │  1.544 s │ 408.80 KiB │   3178 │     128 │
└──────────────┴─────────┴──────────┴──────────┴──────────┴──────────┴────────────┴────────┴─────────┘
```

```
              Shallow water model strong scaling speedup
┌──────────────┬─────────┬─────────┬────────────┬──────────┬─────────┐
│         size │   ranks │ speedup │ efficiency │   memory │  allocs │
├──────────────┼─────────┼─────────┼────────────┼──────────┼─────────┤
│ (8192, 8192) │  (1, 1) │     1.0 │        1.0 │      1.0 │     1.0 │
│ (8192, 8192) │  (1, 2) │ 1.95473 │   0.977364 │ 0.980801 │ 1.18937 │
│ (8192, 8192) │  (1, 4) │ 4.17364 │    1.04341 │ 0.980801 │ 1.18937 │
│ (8192, 8192) │  (1, 8) │ 6.83482 │   0.854352 │ 0.980801 │ 1.18937 │
│ (8192, 8192) │ (1, 16) │ 14.5083 │   0.906771 │ 0.980801 │ 1.18937 │
│ (8192, 8192) │ (1, 32) │  26.233 │   0.819782 │ 0.979154 │ 1.17313 │
└──────────────┴─────────┴─────────┴────────────┴──────────┴─────────┘
```

# System info

```
Oceananigans v0.53.2
Julia Version 1.5.3
Commit 788b2c77c1* (2020-11-09 13:37 UTC)
Platform Info:
  OS: Linux (powerpc64le-unknown-linux-gnu)
  CPU: unknown
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, pwr9)
  GPU: Tesla V100-SXM2-32GB
```

### CODE REVIEW

File: benchmark/strong_scaling_incompressible_model.jl [hunk 1]
```diff
@@ -4,26 +4,48 @@ using Benchmarks
...
+# Run benchmarks
 
 print_system_info()
 
-for r in ranks
-    @info "Benchmarking distributed incompressible model strong scaling [N=($Nx, $Ny, $Nz), ranks=$r]..."
-    julia = Base.julia_cmd()
-    run(`mpiexec -np $r $julia --project strong_scaling_incompressible_model_single.jl $Nx $Ny $Nz`)
-end
+# for R in ranks
```
Comment (francispoulin) [inline benchmark/strong_scaling_incompressible_model.jl:20]: Do we want to keep these comments or remove them?

File: benchmark/strong_scaling_incompressible_model_single.jl [hunk 2]
```diff
@@ -10,36 +10,41 @@ using Benchmarks
...
+Ny = parse(Int, ARGS[3])
+Nz = parse(Int, ARGS[4])
+Rx = parse(Int, ARGS[5])
+Ry = parse(Int, ARGS[6])
+Rz = parse(Int, ARGS[7])
 
-@info "Setting up distributed incompressible model with N=($Nx, $Ny, $Nz) grid points on $R rank(s)..."
+@assert Rx * Ry * Rz == R
+
+@info "Setting up distributed incompressible model with N=($Nx, $Ny, $Nz) grid points and ranks=($Rx, $Ry, $Rz) on rank $local_rank..."
```
Comment (francispoulin) [inline benchmark/strong_scaling_incompressible_model_single.jl:28]: Do we want to say whether it's slab or pencil?  Not even sure if we can but just a thought

File: benchmark/strong_scaling_shallow_water_model_single.jl [hunk 3]
```diff
@@ -15,31 +15,35 @@ MPI.Init()
 local_rank = MPI.Comm_rank(comm)
          R = MPI.Comm_size(comm)
 
-Nx = parse(Int, ARGS[1])
-Ny = parse(Int, ARGS[2])
+decomposition = ARGS[1]
```
Comment (francispoulin) [inline benchmark/strong_scaling_shallow_water_model_single.jl:18]: It occurs to me that there is a lot of overlap between the `incompressible` and `shallow_water` model versions and they could probably be combined, but I don't think we want to do that here.  In the future we can think about how we might dipatch between different models, but not today!

File: src/Distributed/halo_communication.jl [hunk 4]
```diff
@@ -58,12 +58,17 @@ function fill_halo_regions!(c::AbstractArray, bcs, arch::AbstractMultiArchitectu
...
+    x_events_requests = fill_west_and_east_halos!(c, bcs.west, bcs.east, arch, barrier, grid, c_location, args...)
+    y_events_requests = fill_south_and_north_halos!(c, bcs.south, bcs.north, arch, barrier, grid, c_location, args...)
+    z_events_requests = fill_bottom_and_top_halos!(c, bcs.bottom, bcs.top, arch, barrier, grid, c_location, args...)
 
-    events = [west_event, east_event, south_event, north_event, bottom_event, top_event]
-    events = filter(e -> e isa Event, events)
+    events_and_requests = [x_events_requests..., y_events_requests..., z_events_requests...]
+
+    mpi_requests = filter(e -> e isa MPI.Request, events_and_requests) |> Array{MPI.Request}
+    # Length check needed until this PR is merged: https://github.com/JuliaParallel/MPI.jl/pull/458
```
Comment (francispoulin) [inline src/Distributed/halo_communication.jl:68]: thanks for pointing this out!

**Review by francispoulin — APPROVED** [review]
Looks great @ali-ramadhan  and thanks for not only doing all of this but packing it up so nicely!

File: benchmark/strong_scaling_incompressible_model.jl (hunk 1 above)
Comment (ali-ramadhan) [reply benchmark/strong_scaling_incompressible_model.jl:20]: Ah woops, meant to uncomment this. Thanks for catching it!

File: benchmark/strong_scaling_incompressible_model_single.jl (hunk 2 above)
Comment (ali-ramadhan) [reply benchmark/strong_scaling_incompressible_model_single.jl:28]: Ah we can definitely do that!

Comment (francispoulin) [reply benchmark/strong_scaling_incompressible_model_single.jl:28]: Only if you like

File: benchmark/strong_scaling_shallow_water_model_single.jl (hunk 3 above)
Comment (ali-ramadhan) [reply benchmark/strong_scaling_shallow_water_model_single.jl:18]: Yeah they're like 90%+ similar...

I've thought about merging them a bit but feels like we might end up with an overly general (and hard to modify) script.

But would be good to think about merging them in the future.

Comment (francispoulin) [reply benchmark/strong_scaling_shallow_water_model_single.jl:18]: I agree that this is a thougth for the future, but wanted to mention it here so we can mull it over as the code evolves

**Comment by vchuravy** [comment]
Two questions:

> Results seem quite machine-dependent but on Satori for weak scaling I'm seeing >90% efficiency up to 16 ranks then ~85% efficiency on 32 ranks.

Is this one node or multiple nodes? 

Secondly you include a memory measurement. Is this residual memory or memory allocated during a time-step?

**Comment by ali-ramadhan** [comment]
> Is this one node or multiple nodes?

This is all done on one Satori node.

Side note: I couldn't get the benchmark to work with 64 ranks on one Satori node even though it should have 80 ranks per node I think. Also wasn't sure of the best place to immediately benchmark up to 4 ~ 8 nodes (128 ~ 256 ranks?). Probably any old cluster would do but I didn't get a quick `srun` job allocation so I left it for later.

> Secondly you include a memory measurement. Is this residual memory or memory allocated during a time-step?

The memory measurement should be the amount of memory allocated during a single time step as reported by `BenchmarkTools.@benchmark`: generally I think from kernel launches, broadcasts, probably other stuff, etc.

**Comment by francispoulin** [comment]
I guess if we run this on different nodes the efficiency is bound to go down, but that's a problem with the architecture I guess and not something we can change with software.  Unless there is something I don't know about?

**Comment by ali-ramadhan** [comment]
> I guess if we run this on different nodes the efficiency is bound to go down, but that's a problem with the architecture I guess and not something we can change with software. Unless there is something I don't know about?

Yeah I don't have much experience with scaling MPI but I'm guessing single-node scaling depends on your software while multi-node scaling depends more on the cluster/hardware?

**Comment by christophernhill** [comment]
> > Is this one node or multiple nodes?
> 
> This is all done on one Satori node.
> 
@ali-ramadhan would it be hard to do some multi-GPU multi-node fully explicit runs?

**Review by christophernhill — COMMENTED** [review]
@ali-ramadhan aside from mixing the changes to distributed and benchmarks the rest of this is a work of art! Very nice..... 

Do you have thoughts on where we could store more detailed information to try and make things easier for someone else i.e. annoying things like which MPI to use where, CUDA aware MPI foo. Don't think these belong here, but they could be useful to record somewhere?

**Pull request merged**